### PR TITLE
Adds shred command fixing issue #180

### PR DESCRIPTION
**Pull request by JustinTimperio**
One of a few features and fixes I will back port for you from my dev branch

### CODE REVIEW

File: shared/models.go [hunk 1]
```diff
@@ -26,8 +24,10 @@ type Cmd struct {
 }
 
 type Shred struct {
-	Conf shred.Conf
-	Path string
+	Path   string
+	Times  int
+	Zeros  bool
+	Remove bool
```
Comment (lu4p) [inline shared/models.go:30]: Revert this

File: torat_client/rpcapi.go [hunk 2]
```diff
@@ -20,7 +21,19 @@ import (
...
+		Remove: s.Remove,
+	}
+
+	if err := shredconf.Path(s.Path); err != nil {
+		return err
+	}
+	if err := os.RemoveAll(s.Path); err != nil {
+		return err
+	}
+	return nil
```
Comment (lu4p) [inline torat_client/rpcapi.go:36]: Revert this

File: torat_server/client_ishell.go [hunk 3]
```diff
@@ -236,21 +241,20 @@ func (client *activeClient) Reconnect(c *ishell.Context) {
...
+// Shred a remote file
 func (client *activeClient) Shred(c *ishell.Context) {
-	args := shared.Shred{
-		Conf: shred.Conf{
-			Times:  3,
-			Zeros:  true,
-			Remove: true,
-		},
-		Path: strings.Join(c.Args, " "),
-	}
```
Comment (lu4p) [inline torat_server/client_ishell.go:249]: Revert back to this

File: torat_server/client_ishell.go [hunk 4]
```diff
@@ -236,21 +241,20 @@ func (client *activeClient) Reconnect(c *ishell.Context) {
...
-		c.Println(red("[!] Could not shred path:", args.Path, err, r))
+	var s shared.Shred
+	s.Path = strings.Join(c.Args, " ")
+	s.Times = 3
+	s.Zeros = true
+	s.Remove = true
+
+	if err := client.RPC.Call("API.Shred", &s, &void); err != nil {
+		c.Println(red("[!] Could not shred path: ", s.Path))
+		c.Println(red("[!] ", err))
```
Comment (lu4p) [inline torat_server/client_ishell.go:254]: c.Println(red("[!] Could not shred path", s.Path, err))

File: torat_server/client_ishell.go [hunk 5]
```diff
@@ -236,21 +241,20 @@ func (client *activeClient) Reconnect(c *ishell.Context) {
...
+		return
 	}
+	c.Println(green("[+] Finished"))
```
Comment (lu4p) [inline torat_server/client_ishell.go:257]: Shred sucessful

**Review by lu4p — CHANGES_REQUESTED** [review]

**Comment by JustinTimperio** [comment]
@lu4p The way that you have it set up does NOT work. If you would like to revert those changes and test yourself you will see that shred throws errors when the conf is defined the shared struct.

I encourage you to test this yourself if you not believe me.

File: torat_server/client_ishell.go (hunk 4 above)
Comment (JustinTimperio) [reply torat_server/client_ishell.go:254]: Why stack the error in a single line? This is easier to read

Comment (lu4p) [reply torat_server/client_ishell.go:254]: Mostly by convention

File: torat_server/client_ishell.go (hunk 5 above)
Comment (JustinTimperio) [reply torat_server/client_ishell.go:257]: okay, no prob

File: shared/models.go (hunk 1 above)
Comment (JustinTimperio) [reply shared/models.go:30]: Reverting this will result in errors

File: torat_client/rpcapi.go (hunk 2 above)
Comment (JustinTimperio) [reply torat_client/rpcapi.go:36]: Reverting this will result in errors

File: torat_server/client_ishell.go (hunk 3 above)
Comment (JustinTimperio) [reply torat_server/client_ishell.go:249]: Reverting this will result in errors

File: torat_server/client_ishell.go (hunk 4 above)
Comment (JustinTimperio) [reply torat_server/client_ishell.go:254]: I do this elsewhere for readability's sake. Do we want to change that too? Would like to keep consistent.

File: torat_client/rpcapi.go [hunk 6]
```diff
@@ -20,7 +21,19 @@ import (
...
+	shredconf := shred.Conf{
+		Times:  s.Times,
+		Zeros:  s.Zeros,
+		Remove: s.Remove,
+	}
+
+	if err := shredconf.Path(s.Path); err != nil {
+		return err
+	}
+	if err := os.RemoveAll(s.Path); err != nil {
```
Comment (lu4p) [inline torat_client/rpcapi.go:33]: Shred already should do this, if s.Remove is true.

File: torat_server/client_ishell.go [hunk 7]
```diff
@@ -236,21 +241,20 @@ func (client *activeClient) Reconnect(c *ishell.Context) {
...
-		Path: strings.Join(c.Args, " "),
-	}
-	var r error
-	if err := client.RPC.Call("API.Shred", &args, &r); err != nil {
-		c.Println(red("[!] Could not shred path:", args.Path, err, r))
+	var s shared.Shred
+	s.Path = strings.Join(c.Args, " ")
+	s.Times = 3
+	s.Zeros = true
+	s.Remove = true
```
Comment (lu4p) [inline torat_server/client_ishell.go:250]: Please use a struct literal with a shorthand declaration similar to before.

File: torat_server/client_ishell.go (hunk 4 above)
Comment (lu4p) [reply torat_server/client_ishell.go:254]: Yes we want to change that too probably.

File: torat_server/client_ishell.go [hunk 8]
```diff
@@ -50,6 +49,12 @@ func (client activeClient) shellClient() {
 			Help:      "print the content of a file: usage cat <file>",
 			Completer: fileCompleter,
 		},
+		{
+			Name:      "shred",
+			Func:      client.Shred,
+			Help:      "remove a remote file by shreding it: usage shred <file>",
```
Comment (lu4p) [inline torat_server/client_ishell.go:55]: remove traces of a file by overwriting it with random data and zeros and then deleting it.

File: torat_server/client_ishell.go (hunk 5 above)
Comment (lu4p) [reply torat_server/client_ishell.go:257]: Please mark comments as resolved after you have actually fixed them.

**Review by lu4p — CHANGES_REQUESTED** [review]

File: torat_client/rpcapi.go (hunk 6 above)
Comment (JustinTimperio) [reply torat_client/rpcapi.go:33]: In my testing, this was not the case

File: torat_server/client_ishell.go (hunk 8 above)
Comment (JustinTimperio) [reply torat_server/client_ishell.go:55]: okay

File: torat_server/client_ishell.go (hunk 5 above)
Comment (JustinTimperio) [reply torat_server/client_ishell.go:257]: Haven't fixed it yet :)

File: torat_server/client_ishell.go (hunk 7 above)
Comment (JustinTimperio) [reply torat_server/client_ishell.go:250]: Not sure what that means in this context

File: torat_server/client_ishell.go (hunk 4 above)
Comment (JustinTimperio) [reply torat_server/client_ishell.go:254]: okay maybe we just patch this one then I send another PR with the full fix for all the files?

Comment (lu4p) [reply torat_server/client_ishell.go:254]: sure

File: torat_client/rpcapi.go (hunk 6 above)
Comment (lu4p) [reply torat_client/rpcapi.go:33]: Can you raise an upstream issue for shred instead?

Comment (JustinTimperio) [reply torat_client/rpcapi.go:33]: yeah sure. Was just trying to hack it together lol

File: torat_server/client_ishell.go (hunk 7 above)
Comment (lu4p) [reply torat_server/client_ishell.go:250]: See the deleted lines from 242 to 249.

In a struct literal you declare some or all field values of a struct while declaring it, see https://tour.golang.org/moretypes/5.


A shorthand declaration can be used when the type can be infered from the assigned value.

Basically this is the same as
```
var x string = "hello"
```
the shorthand form
```
x := "hello"
```

Comment (JustinTimperio) [reply torat_server/client_ishell.go:250]: ooooooo I see what you mean. You just want to eliminate the var with `:=`. No prob, easy fix here

File: torat_client/rpcapi.go (hunk 6 above)
Comment (JustinTimperio) [reply torat_client/rpcapi.go:33]: @lu4p Do you want me to just remove this and we can fix shred later?

File: torat_client/rpcapi.go [hunk 9]
```diff
@@ -20,7 +21,16 @@ import (
...
+	shredconf := shred.Conf{
+		Times:  s.Times,
+		Zeros:  s.Zeros,
+		Remove: s.Remove,
+	}
+
+	if err := shredconf.Path(s.Path); err != nil {
+		return err
+	}
+	return nil
```
Comment (lu4p) [inline torat_client/rpcapi.go:33]: This can just be
```
return shredconf.Path()
```

File: torat_server/client_ishell.go [hunk 10]
```diff
@@ -236,21 +241,21 @@ func (client *activeClient) Reconnect(c *ishell.Context) {
...
+		Times:  3,
+		Zeros:  true,
+		Remove: true,
 	}
-	var r error
-	if err := client.RPC.Call("API.Shred", &args, &r); err != nil {
-		c.Println(red("[!] Could not shred path:", args.Path, err, r))
+
+	if err := client.RPC.Call("API.Shred", &s, &void); err != nil {
+		c.Println(red("[!] Could not shred path: ", s.Path))
```
Comment (lu4p) [inline torat_server/client_ishell.go:254]: Trailing Space

File: torat_server/client_ishell.go [hunk 11]
```diff
@@ -50,6 +49,12 @@ func (client activeClient) shellClient() {
 			Help:      "print the content of a file: usage cat <file>",
 			Completer: fileCompleter,
 		},
+		{
+			Name:      "shred",
+			Func:      client.Shred,
+			Help:      "remove a path by overwriting it with random data then removing it: usage shred <file>",
```
Comment (lu4p) [inline torat_server/client_ishell.go:55]: `usage shred <path>`

**Review by lu4p — CHANGES_REQUESTED** [review]

File: torat_client/rpcapi.go (hunk 9 above)
Comment (JustinTimperio) [reply torat_client/rpcapi.go:33]: Didn't realize shred with just return nil if no issues.

Comment (lu4p) [reply torat_client/rpcapi.go:33]: Otherwise it wouldn't really make sense to compare an error to nil.

File: torat_server/client_ishell.go (hunk 10 above)
Comment (JustinTimperio) [reply torat_server/client_ishell.go:254]: I'm gonna leave this then patch everything at once with sed

**Pull request merged**